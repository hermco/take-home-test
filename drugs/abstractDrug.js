export default class Drug {
  constructor(name, expiresIn, benefit) {
    const isValidBenefit = 0 <= benefit && benefit <= 50;
    if (!isValidBenefit) {
      throw new Error("a drug benefit value can only be between 0 and 50");
    }

    this.name = name;
    this.expiresIn = expiresIn;
    this.benefit = benefit;
  }

  get hasExpired() {
    return this.expiresIn < 0;
  }

  incrementBenefit() {
    this.benefit = this.benefit + 1;
  }

  decrementBenefit() {
    this.benefit = this.benefit - 1;
  }

  decrementExpiresIn() {
    this.expiresIn = this.expiresIn - 1;
  }

  simulateOneDay() {
    this.decrementExpiresIn();
    this.decrementBenefit();

    if (this.hasExpired) {
      this.decrementBenefit();
    }
  }
}
