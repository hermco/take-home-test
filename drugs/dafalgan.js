import Drug from "./drug";
import config from "../config/config";

export default class Dafalgan extends Drug {
  constructor(expiresIn, benefit) {
    super(config.drugs.DAFALGAN.name, expiresIn, benefit);
  }

  computeBenefitAfterOneday() {
    let i = 0;
    while (i < config.drugs.DAFALGAN.benefitRatioComparedToNormalDrugs) {
      super.computeBenefitAfterOneday();
      i++;
    }
  }
}
