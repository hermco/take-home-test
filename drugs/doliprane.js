import Drug from "./drug";
import config from "../config/config";

export default class Doliprane extends Drug {
  constructor(expiresIn, benefit) {
    super(config.drugs.DOLIPRANE.name, expiresIn, benefit);
  }
}
