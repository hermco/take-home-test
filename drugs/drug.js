import config from "../config/config";

export default class Drug {
  constructor(name, expiresIn, benefit) {
    const isValidBenefit =
      config.MIN_BENEFIT <= benefit && benefit <= config.MAX_BENEFIT;

    if (!isValidBenefit) {
      throw new Error(
        `a drug benefit value can only be between ${config.MIN_BENEFIT} and ${config.MAX_BENEFIT}`
      );
    }

    this.name = name;
    this.expiresIn = expiresIn;
    this.benefit = benefit;
  }

  get hasExpired() {
    return this.expiresIn < 0;
  }

  incrementBenefit() {
    const canIncrementBenefit = this.benefit < config.MAX_BENEFIT;

    if (canIncrementBenefit) {
      this.benefit++;
    }
  }

  decrementBenefit() {
    const canDecrementBenefit = this.benefit > config.MIN_BENEFIT;

    if (canDecrementBenefit) {
      this.benefit--;
    }
  }

  decrementExpiresIn() {
    this.expiresIn--;
  }

  simulateOneDay() {
    this.decrementExpiresIn();
    this.decrementBenefit();

    if (this.hasExpired) {
      this.decrementBenefit();
    }
  }
}
