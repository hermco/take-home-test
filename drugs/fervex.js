import Drug from "./drug";
import config from "../config/config";

export default class Fervex extends Drug {
  _computeBenefitIncrement() {
    if (this.expiresIn < 5) {
      return 3;
    } else if (this.expiresIn < 10) {
      return 2;
    }

    return 1;
  }

  constructor(expiresIn, benefit) {
    super(config.drugNames.FERVEX, expiresIn, benefit);
  }

  simulateOneDay() {
    this.decrementExpiresIn();

    if (this.hasExpired) {
      this.benefit = 0;
    } else {
      const increment = this._computeBenefitIncrement();

      this.incrementBenefit(increment);
    }
  }
}
