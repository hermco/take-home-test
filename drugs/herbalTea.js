import Drug from "./abstractDrug";

export default class HerbalTea extends Drug {
  constructor(expiresIn, benefit) {
    super("Herbal Tea", expiresIn, benefit);
  }

  simulateOneDay() {
    this.decrementExpiresIn();
    this.incrementBenefit();
  }
}
