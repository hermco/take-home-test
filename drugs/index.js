export { default as HerbalTea } from "./herbalTea";
export { default as Doliprane } from "./doliprane";
export { default as Fervex } from "./fervex";
export { default as MagicPill } from "./magicPill";
