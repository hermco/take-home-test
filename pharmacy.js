export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }

  updateBenefitValue() {
    this.drugs.forEach(drug => {
      drug.simulateOneDay();
    });

    return this.drugs;
  }
}
