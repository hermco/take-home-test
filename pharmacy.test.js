import { Pharmacy } from "./pharmacy";
import { Dafalgan, Doliprane, Fervex, HerbalTea, MagicPill } from "./drugs";

describe("Pharmacy", () => {
  it("should simulate days passing by for each of the drugs", () => {
    const drugs = [
      new Doliprane(20, 30),
      new HerbalTea(10, 5),
      new Fervex(5, 40),
      new MagicPill(15, 40),
      new Dafalgan(3, 20)
    ];
    const pharmacy = new Pharmacy(drugs);

    let result;
    for (let elapsedDays = 0; elapsedDays < 30; elapsedDays++) {
      result = pharmacy.updateBenefitValue();
    }

    expect(result).toMatchSnapshot();
  });
});
