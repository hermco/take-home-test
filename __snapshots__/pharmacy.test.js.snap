// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pharmacy should simulate days passing by for each of the drugs 1`] = `
Array [
  Doliprane {
    "benefit": 0,
    "expiresIn": -10,
    "name": "Doliprane",
  },
  HerbalTea {
    "benefit": 50,
    "expiresIn": -20,
    "name": "Herbal Tea",
  },
  Fervex {
    "benefit": 0,
    "expiresIn": -25,
    "name": "Fervex",
  },
  MagicPill {
    "benefit": 40,
    "expiresIn": 15,
    "name": "Magic Pill",
  },
]
`;
